modules/core/hm-mailbox.php

```diff
@@ -164,14 +165,36 @@ public function delete_folder($folder) {
             return;
         }
         if ($this->is_imap()) {
-            $del_folder = prep_folder_name($this->connection, $folder, true);
-            return $this->connection->delete_mailbox($del_folder);
+            $parent_folder = prep_folder_name($this->connection, $folder, true);
+            $this->list_sub_folders[] = $parent_folder;
+            $this->get_recursive_subfolders($parent_folder);
+            $this->list_sub_folders = array_reverse($this->list_sub_folders);
+            foreach ($this->list_sub_folders as $key => $del_folder) {
+                $delete_state = $this->connection->delete_mailbox($del_folder);
+                if ($delete_state) {
+                    unset($this->list_sub_folders[$key]);
+                }
+            }
+            if (empty($this->list_sub_folders)) {
+                return true;
+            } else {
+                return false;
+            }
         } else {
             $del_folder = decode_folder_str($folder);
             return $this->connection->delete_folder($del_folder);
         }
     }
 
+    public function get_recursive_subfolders($parentFolder) {
```

Comment:
hm-mailbox is a bridge towards different connections (protocols) for accessing mail. What you do with the changes here is imap-specific, so please move it to hm-imap file where it belongs.